aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,647 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/Parallel.h>
+#include <ATen/NamedTensorUtils.h>
+#include <map>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename index_t>
+index_t get_nvalues(const std::vector<index_t>& sizes, const int64_t sparse_dim) {
+  /* Return the number of entries in the dense part of a sparse tensor.
+
+     `sizes` is a vector of sparse tensor dimensions.
+     `sparse_dim` is the dimension of the sparse part of a sparse tensor.
+   */
+  index_t nvalues = 1;
+  for (auto it = sizes.begin() + sparse_dim; it != sizes.end(); ++it) {
+    nvalues *= *it;
+  }
+  return nvalues;
+}
+
+template <typename index_t>
+std::vector<index_t> get_offsets(const Tensor& indices, const std::vector<index_t>& sizes, const int64_t dim) {
+  /*
+    Given the indices of a sparse tensor, return a vector of offsets
+    for the entries in the equivalent dense tensor:
+
+      If
+        offsets = get_offsets(A._indices(), A.sizes(), -1)
+        data = A.to_dense().resize((nnz,))
+      then
+        data[offsets[n]] == A._values()[n]
+
+    `indices` must be a contiguous 2-d tensor with index_t entries.
+    `sizes` must be a vector with at least ndim entries.
+
+    `dim` is an integer. When >= 0 and < ndim, the indices of all
+    entries in the given dimension will be mapped to the index of the
+    first entry before computing the offset. Otherwise, the value is
+    ignored.
+
+    For example, consider a sparse tensor
+
+      11 ** ** 14 15
+      ** 22 ** 24 **
+
+    with
+
+      indices = [[0, 0, 0, 1, 1],
+                 [0, 3, 4, 1, 3]]
+
+    then
+
+      get_offsets(indices, (2, 5), -1) -> [0, 3, 4, 6, 8]
+      get_offsets(indices, (2, 5), 0) -> [0, 3, 4, 1, 3]
+      get_offsets(indices, (2, 5), 1) -> [0, 0, 0, 5, 5]
+
+  */
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  std::vector<index_t> offsets(nnz);
+  std::vector<index_t> strides(ndim, 1);
+  auto indices_accessor = indices.accessor<index_t, 2>();
+
+  if (ndim > 1) {
+    for (int64_t i=ndim - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * sizes[i + 1];
+    }
+  }
+
+  for (int64_t i=0; i < nnz; i++) {
+    index_t acc = 0;
+    for (int64_t j=0; j < ndim; j++) {
+      auto indices_row = indices_accessor[j];
+      auto stride = strides[j];
+      if (j != dim) {
+        acc += stride * indices_row[i];
+      }
+    }
+    offsets[i] = acc;
+  }
+
+  return offsets;
+}
+
+template <typename index_t>
+std::vector<std::vector<index_t>> get_pools(const Tensor& indices, const std::vector<index_t>& sizes, const int64_t dim) {
```

Comment:
I guess this returns a vector of vectors because it's ragged?

Reply:
yes, exactly.